### PR TITLE
Convert Html Docs to Sphinx. Add to readthedocs

### LINKED ISSUE #1204: Convert SWIG documentation to RST/Sphinx, add to ReadTheDocs
SWIG's documentation currently relies on hand-crafted HTML code, supplemented by a few decades-old Python scripts. Writing HTML by hand is cumbersome and provides another obstacle to improving the documentation when new features are added.

I'd like to convert SWIG's documentation to RST. Its text form is human-readable and -writable, making it easier to modify and extend. Generating HTML with Sphinx uses pygments for syntax highlighting, which [supports most/all existing target languages](http://pygments.org/languages/) and [SWIG as well](http://pygments.org/docs/lexers/#pygments.lexers.c_like.SwigLexer).

Having SWIG on `readthedocs.org` might also solve #428 , since there's always one "latest" version of the documentation available, and the ability for pages to be broken down into subsections improves their ability to be linked from google.

I'm willing to take lead on this -- I have experience with Sphinx, RTD, Pandoc, and using Python to automate whatever tweaks might be necessary to get them all working together properly.

### PR DESCRIPTION
**Pull request by chsasank**
Fix swig/swig#1204 

Following the dicussion in swig/swig#1204, I've created sphinx and hosted them on readthedocs

https://swig.readthedocs.io/en/latest/index.html

image

I've taken some liberties:

1. Removed numbering, because they're hard to maintain.
2. Removed index.html and Section.html in Manual/. Instead link them in root index.rst
3. Links are working, but the way current linking works is not the best practice in RST

There's still decent amount of documentation that needs to be ported. IMO, whole swig.org website can be hosted in readthedocs. Please give feedback and I'll port and create the docs in readthedocs.

Also:

Read the docs could be branded to swig's branding (colors etc). See https://simpleitk.readthedocs.io/en/master/

### CODE REVIEW

**Comment by tleonhardt** [comment]
That’s a huge step forward in terms of maintainability compared to hand-coded HTML.  Nice work!

**Comment by vadz** [comment]
Thanks a lot for working on this!

FWIW I also think that it's great and while I'd prefer Markdown to RST, the latter is still incomparably better than raw HTML and docs appearance on RTD is day and night compared to the current one. I hope @wsfulton can be convinced to integrate this soon, before it gets out of date due to the ongoing changes in the HTML docs.

**Comment by chsasank** [comment]
You might be also interested in this idea for swig's website: https://chsasank.github.io/swig_www/
Completely markdown based, zero build, website hosted on github pages: https://github.com/chsasank/swig_www/

image

**Comment by vadz** [comment]
This is amazing, but I wonder how exactly does this work? The "zero build" part looks like "black magic" to me...

**Comment by chsasank** [comment]
Github allows 'remote_theme': https://github.blog/2017-11-29-use-any-theme-with-github-pages/
So I have used just_the_docs theme: https://pmarsceill.github.io/just-the-docs/#quick-start-use-as-a-github-pages-remote-theme

**Comment by wsfulton** [comment]
@chsasank, my initial impression is this is very good and I'll be taking a deeper dive soon with a view to switching over. 

I was wondering how you converted the hand written html files to rst format and how error prone this is?

I skimmed the first section in the Java chapter and found there is some missing text in the "Java specific options" table in http://swig.org/Doc4.0/Java.html#Java_commandline compared with https://swig.readthedocs.io/en/latest/Manual/Java.html#additional-commandline-options. I'm not going to compare the entire doc to check everything! So, how can we be sure there aren't other transcribing errors?

I see readthedocs supports https, so putting the entire website there could solve https://github.com/swig/www/issues/8.

Is it possible for readthedocs to host the pages, but still 

Will this also solve google-ability? see #428.

File: SphinxDocs/post_process.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import re
+import glob
+
+number_regex = re.compile(r'^\d[\.\d]+ ', flags=re.M)
+for f in glob.glob('source/Manual/*.rst'):
```
Comment (chsasank) [inline SphinxDocs/post_process.py:5]: > html files to rst format and how error prone this is?

Although pandoc is not foolproof, it's pretty good. With basic regex, I could easily modify and make it better.

File: SphinxDocs/import.sh [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+mkdir -p $sphinx_src_dir
+
+for html in $root_html_doc_dir/*.html
+do
+    echo converting $html
+    html_base_name=$(basename $html)
+    rst_base_name=${html_base_name/html/rst}
+    rst=$sphinx_src_dir/$rst_base_name
+    pandoc -o $rst $html
```
Comment (chsasank) [inline SphinxDocs/import.sh:16]: > I was wondering how you converted the hand written html files to rst format and how error prone this is?

I've used [pandoc](https://pandoc.org/). I've convert html documentation files to RST and html website to MD using pandoc.

**Comment by chsasank** [comment]
> I see readthedocs supports https, so putting the entire website there could solve swig/www#8.

Yes. readthedocs supports https.

> Is it possible for readthedocs to host the pages, but still

You can use CNAME dns from https://docs.swig.org to https://swig.readthedocs.io. Hosting and building is completely managed by readthedocs.

> Will this also solve google-ability? see #428.

Yes, using readthedocs (and justthedocs) significantly improves google-ability. URL for latest documentation always remains samg. Eg. [https://swig.readthedocs.io/en/**latest**/Manual/CPlusPlus17.html](https://swig.readthedocs.io/en/latest/Manual/CPlusPlus17.html). You can also maintain multiple versions at the same time in one place.

image

Although the move to readthedocs might not be super smooth and requires _some_ manual work, I think it is worth in long term. It's very easy to update and maintain documentation in RST than manual HTML.

I will help what I can to make this process smooth.

**Comment by chsasank** [comment]
> I skimmed the first section in the Java chapter and found there is some missing text in the "Java specific options" table in http://swig.org/Doc4.0/Java.html#Java_commandline compared with https://swig.readthedocs.io/en/latest/Manual/Java.html#additional-commandline-options. I'm not going to compare the entire doc to check everything! So, how can we be sure there aren't other transcribing errors?

Pandoc is error prone as you observed. We need a decent way to check diffs. May be we can convert both html and convert rst files to text and do diff on text files. After looking at this diff, export errors will be obvious

File: SphinxDocs/post_process.py (hunk 1 above)
Comment (chsasank) [reply SphinxDocs/post_process.py:5]: Note: After automatic conversion, I did some manual editing too myself.

**Comment by chsasank** [comment]
Hi,

I've written better script at doing the diffs
```
set -e
html2text --body-width=0 Doc/Manual/Java.html > /tmp/java1.txt 
rst2html.py SphinxDocs/source/Manual/Java.rst | html2text --body-width=0 > /tmp/java2.txt
git diff --word-diff /tmp/java1.txt /tmp/java2.txt | grep '{+\|\[-' | grep -v '^\[-#' > /tmp/java.diff
```

This one picks up the above diff.
[Uploading java.diff.txt…]()

**Comment by wsfulton** [comment]
I think we do need to do a thorough diff check. @chsasank, can you/have you done it for all the html files?

> You can use CNAME dns from https://docs.swig.org to https://swig.readthedocs.io. Hosting and building is completely managed by readthedocs.

So will that allow the entire site to appear at https://swig.org? And if so, will it be the only place to get the content, that is, the content will not be available from swig.readthedocs.io? We don't really want a duplicate site!

**Comment by chsasank** [comment]
I have not done a diff check. I unfortunately don't have expertise (or time) to go through the diff. I can certainly generate the above diff for someone to go through and verify.

Yes. There can be two websites - swig.org and docs.swig.org for landing and documentation respectively. There'll not be any 'duplicates'.

**Comment by wsfulton** [comment]
@chsasank I've taken your branch and created a `sphinxdocs` branch from it, see https://github.com/swig/swig/commits/sphinxdocs. I've taken your script to do the diffs and improved it. It can be run like this:
```
~/swig/github/swig/SphinxDocs (sphinxdocs=)$ ./docdiffer.sh -d=gvimdiff Java
Running: gvimdiff /tmp/Java-old.txt /tmp/Java-new.txt
```
to diff the output comparing the old Java.html against the new Java.rst file as text diffs using a visual diff tool (gvimdiff). I updated the Java.rst file and now the diffs are inconsequential. There are many other files to inspect and fix. If you or anyone else can help, please let me know, otherwise it will take me a while to get through them and hold up other dev work. Tools required to get going are https://github.com/swig/swig/blob/38c9ce7ffd4eac055d8aa34fc0b21df21d020d4b/SphinxDocs/docdiffer.sh#L7-L8

**Comment by wsfulton** [comment]
btw, how can I view the html files on my local machine but formatted the same way that they are on readthedocs?

**Comment by chsasank** [comment]
> If you or anyone else can help, please let me know, otherwise it will take me a while to get through them and hold up other dev work. 

Unfortunately it's hard for me to do the diff check because I neither have expertise to review swig's docs nor time to do it. Sorry :(. May be swig veterans can help?

> btw, how can I view the html files on my local machine but formatted the same way that they are on readthedocs?

You can open _build/html/index.html in some browser and it should work just fine!

**Comment by wsfulton** [comment]
> I will help what I can to make this process smooth.

I need some help!

We're missing usage/installation instructions. I have installed the following:
```
sudo apt install python3-sphinx python3-sphinx-rtd-theme pandoc python3-docutils html2text
```
and run:
```
~/swig/github/swig/SphinxDocs $ make html
Running Sphinx v1.6.7
WARNING: extension 'sphinx_rtd_theme' has no setup() function; is it really a Sphinx extension module?
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 57 source files that are out of date
updating environment: 57 added, 0 changed, 0 removed
reading sources... [100%] index                                                                       
/home/william/swig/github/swig/SphinxDocs/source/Devel/internals.rst:1029: WARNING: Unexpected section title or transition.

--------------
/home/william/swig/github/swig/SphinxDocs/source/Devel/scanner.rst:239: WARNING: Inline emphasis start-string without end-string.
```
a screenful of warnings follows. Do you get these warnings? They are serious problems in the patch from what I can make out.
I get an error and no html is generated.
```
preparing documents... 
Theme error:
unsupported theme option 'style_external_links' given
Makefile:24: recipe for target 'html' failed
```
If it is this hard to generate the docs, then no-one is going to edit and contribute. We need robust instructions / tools so that it just works otherwise all this is a retrograde step.

>> If you or anyone else can help, please let me know, otherwise it will take me a while to get through them and hold up other dev work.

> Unfortunately it's hard for me to do the diff check because I neither have expertise to review swig's docs nor time to do it. Sorry :(. May be swig veterans can help?

I've fixed a handful of .rst files and the required skills have nothing to do with SWIG. They require rst knowledge which I suspect veterans do not have. There are some problems that I couldn't figure out, eg how do you display the double backslashes in the tables in Doxygen.rst? Given the push to use rst, I presume you have the knowledge to fix the input rst, something few of us have.

**Comment by chsasank** [comment]
> We're missing usage/installation instructions. I have installed the following:

I'll write a README.md in SphinxDocs directly! My bad for not documenting them. 

> Do you get these warnings? They are serious problems in the patch from what I can make out.

These are not a big deal AFAIK, but I'll take a look.

>  I get an error and no html is generated.

This is of course more serious. It's a missed dependency. I'll write instructions in README.md

> If it is this hard to generate the docs, then no-one is going to edit and contribute

Ideally it should not be this hard. But I get you - I should have documented well!

> We need robust instructions / tools so that it just works otherwise all this is a retrograde step.

You needn't worry! Nobody from swig ever asked me to write instructions until now. Sphinx is fairly easy to thing to manage. Projects - small to big - even, linux kernel moved to sphinx!

>  They require rst knowledge which I suspect veterans do not have.

I didn't understand that you wanted help with rst - certainly, happy to help there! I do have fair bit of experience working with rst.

> There are some problems that I couldn't figure out, eg how do you display the double backslashes in the tables in Doxygen.rst?

I understand that there's some problem on doxygen.rst. I'll check that out. But otherwise, can we make a list of things where I can help on? How would I know your problems if you don't tell me :)

Overall, I think the communication b/w us could be better and it'll make things smoother. I have talked to maintainer of [Lark](https://github.com/lark-parser/lark) on a chat platform (Gitter) and [moved their documentation](https://github.com/lark-parser/lark/pull/655) to sphinx and readthedocs in a day or two. I have even supported them [going live](https://github.com/lark-parser/lark/pull/655#issuecomment-674833942).

Given I worked with sphinx before, it doesn't take much time for me to fix all the above problems. It'd help me if I can do the whole migration thing in one sitting, like I have done for Lark, as opposed to multiple small sittings spread over months. We all know how we lose the touch on code as we stop working on it.

I estimate I should be able to complete the migration in this weekend. But can you please promise me communication and support from you? I suppose our timezones certainly doesn't help with the communication. If this weekend is not cool, we can do this next weekend. My weekdays tend to be quite packed and so, I can't really justify working on this at my work.

I'll wait for your reply before I start working on the above points.

**Comment by wsfulton** [comment]
>> Do you get these warnings? They are serious problems in the patch from what I can make out.

> These are not a big deal AFAIK, but I'll take a look.

Well what I can see in the generated html is rst warnings. Also missing text and wrong output from rst syntax gone wrong. All these need fixing in the rst.

The list of problems is too long to list here. But they are clearly visible using your idea of diffing the text output. I took your suggestion further so that it can fire up a diff tool of your choice. You should be able to use your favourite graphical diff tool, things like meld or gvimdiff to make it easy to find these problems. If this isn't clear from what I posted earlier, let me know.

I've just committed some of the fixes I've managed to work out so far, so please base any further changes off the sphinxdocs branch. If you can do this in one sitting that would be great. I should be able to respond fairly quickly this weekend.

**Comment by wsfulton** [comment]
There are various missing links as well btw, such CPlusPlus11.rst
```rst
``%constant`` can then be used for any callback function as described in
`Pointers to functions and callbacks <SWIG.html#SWIG_nn30>`__.
```
I've just pasted the code rather than show a permalink as I can't figure out how to get a permalink to a .rst file in Github as Github turns rst files into html rather than show you the source

I run a link checker on the html docs every now and again. See the linkchecker1, linkchecker2, linkchecker3 targets in https://github.com/swig/swig/blob/edbf6240bae89e76f91012c6733ac9f987413d96/Doc/Manual/Makefile#L77-L97

We'll need to do some sort of equivalent or find something in sphinx to provide an equivalent.

**Comment by chsasank** [comment]
Added instructions for building and writing docs. Can you go through/follow the instructions so that we both can complete the work together! Let me know if any instructions are not clear/incomplete?

Hacking away at other things.

**Comment by chsasank** [comment]
I understand what you mean missing links. Basically #id links are not working as expected. This is the sphinx's equivalent: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#ref-role

**Comment by chsasank** [comment]
> how do you display the double backslashes in the tables in Doxygen.rst?

Well you did figure something, right?

```
    * - ``\\``
      - prints \\ char
```

**Comment by chsasank** [comment]
I have checked until Doxygen.rst in alphabetical order. Can you check next few files and ask me if you have any questions? Build/dev instructions are in readme.

**Comment by wsfulton** [comment]
The double backslash solution I put in is still no good, the generated html is:
```html
<tr><td><tt class="docutils literal">\\</tt></td>
```
It needs to be:
```html
<tr><td>\\</td>
```
So that the formatting matches up with the rest of the table.

**Comment by wsfulton** [comment]
My earlier commits have fixed the files in alphabetical order up to Doxygen.rst (except the linking problems), so they are nearly okay now. How about you work backwards alphabetically and I'll work forwards?

**Comment by chsasank** [comment]
Lol, that's why there were fewer things to fix. Sure I will work backwards alphabetically. 

Well I too don't know how to get // working. We'll look at it later I guess

**Comment by geographika** [comment]
For a double backslash I think you just have to escape it twice:

`This is a path \\\\`

From rst.ninjs.org this outputs 

`This is a path \\`

**Comment by chsasank** [comment]
Tried it. It doesn’t seem to work in sphinx.

**Comment by wsfulton** [comment]
@chsasank , how far did you get? I got to Extensions.rst so far.

**Comment by chsasank** [comment]
Sorry but I got hooked up into some other work over Sunday. I didn't get far. 

I'll try to complete this in my free time now.

**Comment by wsfulton** [comment]
This work is going to be redundant and will end up being thrown away if we don't get it finished soon. Are there any other Sphinx guru's able to lend a bit of time checking and fixing the restructured text markup?

Any thoughts perhaps on whether this is not the right way forward for future SWIG documentation?

**Comment by chsasank** [comment]
If you can ask me specific questions, I can certainly help you out!

**Comment by vadz** [comment]
Sorry, it's not really clear to me what is/are the blocker(s) right now. Could we perhaps have a list of checkboxes in this PR to see what remains to be done at a glance?

FWIW personally I think it would be better to merge this right now, even if there are some (hopefully temporary) regressions in the docs, than not to merge it at all because I'm pretty sure that the gains in documentation quantity and quality in the future thanks to using ReST rather HTML will more than compensate for them.

**Comment by wsfulton** [comment]
A high level view of the blockers are:

* Broken links.
* Missing text.
* Formatting errors.

@chsasank and I have worked together to provide tools to find these problems. If someone wants to itemise them one by one please go ahead, but I think it will be more a lot more efficient to spend the time fixing them instead of itemising them in great detail as many of the problems are easy ReST fixes.

If no-one is prepared to help fix these problems, then this is a step backwards and not worth considering the change as it isn't good enough to publish and I have no intention of maintaining two sources of documentation. If using ReST will make the docs more user friendly and enable more users to contribute easily, then I'd hope this would be proved by SWIG users fixing these problems now!

**Comment by vadz** [comment]
FWIW I've quickly looked at [Doxygen chapter](https://swig.readthedocs.io/en/latest/Manual/Doxygen.html) and it seems good enough to publish to me. At a glance, I couldn't see anything broken/missing.

Also, the point I was trying to make was that it's much easier to modify or add new documentations to ReST files, not that it's easy to find differences between the old HTML and new ReST, which is obviously not easy at all. You can hardly take this as a proof that switching to ReST is not worth it!

**Comment by chsasank** [comment]
Doxygen looks cool because @wsfulton has done most of the work on it :). Checkout in reverse alphabetical order!

**Comment by chsasank** [comment]
@manitejav has volunteered to help us with this project - thanks a lot Mani Teja! As you can see in the above rather long discussion, we've used pandoc to convert html based swig docs to rst and sphinx. However, pandoc is not perfect and there are some inconsistencies to be fixed. @wsfulton and I wrote a [script](https://github.com/chsasank/swig/blob/master/SphinxDocs/docdiffer.sh) to automatically detect these inconsistencies. We have used this script to fix inconsistencies until `Extension.rst` in alphabetical order. Can you please help us by fixing rest of the inconsistencies in reverse alphabetical order. Thanks again!

I have given you write access to my fork - [chsasank/swig](https://github.com/chsasank/swig). Push your commits to the master branch there. This PR will get updated automatically.

**Comment by wsfulton** [comment]
Hi @manitejav. Thanks for the updates. One thing, could you remove sectiontoc you've added into each of the .rst files. The table of contents appears on the rhs in readthedocs and so is not necessary. In your commits, please summarise the .rst file you have changed 'updated rst' doesn't help much as it does not say which file you are modifying.

**Comment by wsfulton** [comment]
f7456cc  removes some corrections I put in. Please revert these. Can you rebase your commits too to fix these problems.

**Comment by wsfulton** [comment]
fafe621 - @chsasank we don't need this now right?

**Comment by chsasank** [comment]
> fafe621 - @chsasank we don't need this now right?

Indeed. Contents are auto generated!

**Comment by wsfulton** [comment]
I've already done Java.rst, so Java changes in 13c20b2 are not needed. You've used a really hard to maintain table format, the `list-table` format I used is much better. Not sure why did that, nor the other formatting changes which I improved.

**Comment by manitejav** [comment]
> I've already done Java.rst, so Java changes in [13c20b2](https://github.com/swig/swig/commit/13c20b2e2539e9e8634cd0181b70e9a91761b91f) are not needed. You've used a really hard to maintain table format, the `list-table` format I used is much better. Not sure why did that, nor the other formatting changes which I improved.

@wsfulton Will make necessary changes, rebase and push it.

**Comment by wsfulton** [comment]
If pushing a lot, please use `[skip-ci]` in your commits to avoid the massive amount of CPU time needed for testing when nothing is going to change. Thanks.

**Review by wsfulton — COMMENTED** [review]
Please document this new Warnings_script.py with purpose and usage instructions and give it a name that indicates its purpose. Perhaps it is just a temporary thing and shouldn't have been committed?

**Comment by manitejav** [comment]
I have made necessary changes to rst files and the lastest version is at https://swig.readthedocs.io/en/latest/index.html
@wsfulton can you please review the changes.

Also are you interested in this. https://github.com/swig/swig/pull/1844#issuecomment-663332068

File: SphinxDocs/source/Manual/D.rst [hunk 3]
```diff
@@ -333,8 +335,8 @@ When generating D wrappers, a few additional macros are available:
 
 ``$dcall`` and ``$winput`` (C#: ``$cscall``, ``$iminput``)
    These variables are used in the director-specific typemaps
-   ```ddirectorin`` <D.html#D_ddirectorinout>`__ and
-   ```ddirectorout`` <D.html#D_ddirectorinout>`__. They are more or less
+   :ref:`ddirectorin <D_ddirectorinout>` and
+   :ref:`ddirectorout <D_ddirectorinout>`. They are more or less
```
Comment (wsfulton) [inline SphinxDocs/source/Manual/D.rst:339]: When I run docdiffer.sh, this change results in:
```
...
+ rst2html.py ../SphinxDocs/source/Manual/D.rst
../SphinxDocs/source/Manual/D.rst:339: (ERROR/3) Unknown interpreted text role "ref".
../SphinxDocs/source/Manual/D.rst:339: (ERROR/3) Unknown interpreted text role "ref".
```
Firstly, I don't really know what this is about as the link seems to work. Secondly, I don't see this error when I run `make html`. Why not?

Comment (manitejav) [reply SphinxDocs/source/Manual/D.rst]: Using an extra space solves the problem. I pushed the changes.
In `docdiffer.sh` you are using `rst2html` whereas in make you are using `sphix`. Both interpret the input differently. That's the reason `make.html` didn't give any error.

Comment (wsfulton) [reply SphinxDocs/source/Manual/D.rst]: now the output of rst2html is wrong, surely this should work properly. Why did you use `:ref:` and not the normal linking syntax?

Comment (wsfulton) [reply SphinxDocs/source/Manual/D.rst]: @manitejav ???

**Comment by wsfulton** [comment]
How can we fix the formatting in the Function table in this section: https://swig.readthedocs.io/en/latest/Manual/SWIG.html#advanced-renaming-support so that it is the same as the original: http://swig.org/Doc4.0/SWIG.html#SWIG_advanced_renaming ?

**Comment by wsfulton** [comment]
@manitejav, thanks you've done the majority of the fixes. I've found some more using the docdiffer.sh script when checking. I've now checked the Manual directory and it has no more textual regressions. The https://github.com/swig/swig/commits/sphinxdocs branch should now be used for the latest.

I am not happy with the formatting regression, see prev comments, so am looking for fixes there.
I also don't have answers to my link issues, see prev comments. Next steps are to check for linking regressions, then we should be able to merge to master. I'll adapt our previous link checking tools in Doc/Manual/Makefile next and report back any link problems.

**Comment by wsfulton** [comment]
> Also are you interested in this. #1844 (comment)

I am. Let's get the Documentation sorted first. I suggest a separate pull request for this conversion of the main website to markdown. BTW, why not choose rst instead?

**Comment by manitejav** [comment]
> How can we fix the formatting in the Function table in this section: https://swig.readthedocs.io/en/latest/Manual/SWIG.html#advanced-renaming-support so that it is the same as the original: http://swig.org/Doc4.0/SWIG.html#SWIG_advanced_renaming ?

Will look into it and update.

File: SphinxDocs/source/Manual/D.rst (hunk 3 above)
Comment (manitejav) [reply SphinxDocs/source/Manual/D.rst]: @wsfulton For internal linking you need :ref: https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#cross-referencing-arbitrary-locations.

**Comment by wsfulton** [comment]
The internal links can be checked using:
```
sudo apt install w3c-linkchecker
cd SphinxDocs/build/html/Manual/
checklink -q -r -X "^http.*" ./index.html
```
This checks the links in the generated html files that are not links to external sites. It produces a report showing hundreds of broken internal links. It seems that the vast majority of the internal links within the SWIG documentation are now broken.

I'm sure these can be fixed with some effort. Anyone willing to do this? Before doing so, I'd like to know what approach will be taken as there are a couple of ways to deal with links within rst documents.

**Comment by wsfulton** [comment]
@chsasank this last comment should have been addressed to you instead. Sorry for mixup, but can you please reply?

**Comment by chsasank** [comment]
Hi all,

Sorry for late reply. I have picked this up back in 2020 as summer project and didn't look at it in 2021.

Internal links are the hardest to fix if we use basic pandoc based scripts. We need to find some pattern in raw pandoc's conversion and use that to script to something that makes sense for sphinx.

**Comment by chsasank** [comment]
I just checked some of the rst and the documentation website, the internal links seems to be working _somewhat_

For example, see SwigPlus page [here](https://swig.readthedocs.io/en/latest/Manual/SWIGPlus.html)

```
This chapter describes SWIG's support for wrapping C++. It is mostly
concerned about C++ as defined by the C++ 98 and 03 standards. For
additions to the original C++ standard, please read the `SWIG and
C++11 <CPlusPlus11.html#CPlusPlus11>`__, `SWIG and
C++14 <CPlusPlus14.html#CPlusPlus14>`__ and `SWIG and
C++17 <CPlusPlus17.html#CPlusPlus17>`__ chapters. `SWIG and
C++20 <CPlusPlus20.html#CPlusPlus20>`__ chapters. As a prerequisite, you
should first read the chapter `SWIG Basics <SWIG.html#SWIG>`__ to see
how SWIG wraps ISO C. Support for C++ builds upon ISO C wrapping and
that material will be useful in understanding this chapter.
```

All the links are working fine. But the issue is the `#` part actually. Basically chapter names are not being mapped correctly. I have a solution:

1. Make a list of all such links in the documentation with #. 
2. Make a list of all chapter names from sphinx. Example link: https://swig.readthedocs.io/en/latest/Manual/Preprocessor.html#swig-macros
3. Do some fuzzy mapping
4. Script the changes

**Comment by wsfulton** [comment]
Yeah, the link to the page is okay, but that is a minor part of the link. The links need to go to the correct position within the page (it's ignoring the link '#' as you pointed out). All the ones you mentioned go to the top of the page, they were intended to go to the top of the page. If you look for `feature flag` on the https://swig.readthedocs.io/en/latest/Manual/SWIGPlus.html page, the https://swig.readthedocs.io/en/latest/Manual/Customization.html#Customization_feature_flags link does not work correctly. Compare with the official docs http://swig.org/Doc4.0/SWIGPlus.html and that equivalent link does work to http://swig.org/Doc4.0/Customization.html#Customization_feature_flags.

I look forward to your 4 point plan working, even though I don't fully follow it. 

https://swig.readthedocs.io comes up really high on Google search and these docs are quite broken, are just test quality and far too public. Can you please remove them until a satisfactory solution is found?

**Comment by wsfulton** [comment]
Closing with a heavy heart as there seems to be no solution to the fundamental requirement of a documentation system of supporting links within a document. All this work has gone to waste and this branch is now very stale. What a shame.

I suggest alternative solutions are discussed on #1204.

**Comment by LecrisUT** [comment]
Can we re-open this? Links to headers is supported in sphinx, it's just that the naming convention needs to be fixed

**Comment by wsfulton** [comment]
The links don't work. The technical details to make them work are very much missing. If someone provides a patch or full details with examples to make them work, then it might be considered, but until then this approach is a dead end. However, the converted docs are now very out of date now and so the full conversion from the latest in master would need redoing. Best to follow #1204 for future solutions.

**Comment by LecrisUT** [comment]
@wsfulton I did not do any fixes to the swig, but just to illustrate that the links work:
[`Manual/Customization.html#feature-flags`](https://swig-experimental.readthedocs.io/en/latest/Manual/Customization.html#feature-flags). Note that the naming convention is changed from [`#Customization_feature_flags`](http://swig.org/Doc4.0/Customization.html#Customization_feature_flags) to [`#feature-flags`](https://swig-experimental.readthedocs.io/en/latest/Manual/Customization.html#feature-flags) which is derived as header name to lowercase + spaces to `-`. It would be possible to regex the links to fix them since the old model also has a predictable format, but also we can change the link names by adding a prefactor [`(Customization_feature_flags)=`](https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html#creating-explicit-targets).

I would otherwise strongly recommend using sphinx as the backend, particularly due to the `doxygen` support, and the cross-link to other projects which are also generally sphinx based.

**Comment by wsfulton** [comment]
Hi @LecrisUT. Thanks and good to know. Unfortunately converting the latest doc updates is going to be time consuming and probably requires starting again. It took a month of my spare time to do this previously and was very laborious with lots of manual editing. So I can't see this happening anytime soon. There is also the markdown approach used in #1195 which could be used. I'm not sure which is better now. 

Meanwhile, https://swig-experimental.readthedocs.io/en/latest/Manual/Preface.html#introduction is mistakenly being found and used by SWIG users and assumed to be the official docs. See #2923. In order to avoid any confusion and time wasting by users and developers alike going forwards, could you please remove these docs from readthedocs.

**Comment by LecrisUT** [comment]
Firstly it would be appreciated if it didn't take 10 months to reply. Secondly, the tooling have constantly evolved since the original attempts were proposed here which would, so it wouldn't take months of *your* time, especially as other people are volunteering their time if you would just take it.

Seriously, please interact with the community better. We all believe in the potential that this project can have, and we want to volunteer our time to grow it. We each have our strengths, and it is ok to rely on other people to review, and share their expertise, so that it is not a chore and time-sink for you and the core developers.

Detecting the broken links is a trivial task since there is `linkcheck`, and it would not take more than one evening of the person volunteering to implement this.

> There is also the markdown approach used in https://github.com/swig/swig/pull/1195 which could be used. I'm not sure which is better now.

It is the same approach that is proposed here, same usage of pandoc. The either format is accepted by sphinx and rtd. The only thing required is to *make a decision* and pull the trigger. Converting from rst to and from markdown is 100% compatible if it's decided to change.

> In order to avoid any confusion and time wasting by users and developers alike going forwards, could you please remove these docs from readthedocs.

Question is, is swig serious about modernizing the documentation? I can make it only run on PRs so that it is hidden, and spend a weekend to rebase everything on the latest documentation and get this done. But if the core developers are not serious about this, then making it clear in #1204 that this approach will not be considered is preferred.

**Comment by wsfulton** [comment]
@LecrisUT would you kindly take down https://swig-experimental.readthedocs.io/en/latest/Manual/Preface.html#introduction as requested or is there a reason it is not being done?

Please discuss alternative documentation systems in an #1204 instead of in this closed ticket.

**Comment by LecrisUT** [comment]
I would reply in kind, would you kindly not take 3-10 months to reply to a comment? Feel free to copy and quote reply my comments and ping me where you think it's appropriate. I have tried to discuss topics several times, but each time I am met with pure silence, and if this is the _only_ way to maintain a conversation even if it's with 3-10 months delays, so be it.

**Comment by wsfulton** [comment]
If it isn't clear, this pull request is closed, my fingers were burnt on this, I avoid it as it was my worst decision as SWIG maintainer to spend so long on trying to get it over the line. Please please discuss documentation on an open thread as requested. My only answer I have to your questions is forcing dashes in a link is a non-starter for me personally as you can't find them using whole word searching in editors (so is a big step backwards for developers editing the docs). 

While I'm not the only person that needs convincing about putting in a new documentation system, I suggest putting forward a demo in a new pull request to get some fresh impetus. I suggest it shows just two converted chapters (with cross links)  discussing how it works, pros/cons etc for everyone to comment on. Removing https://swig-experimental.readthedocs.io is a priority please and then replace with a two chapter sample would be ideal.

**Pull request closed without merging**